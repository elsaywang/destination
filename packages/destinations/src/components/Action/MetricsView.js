import React from 'react';
import PropTypes from 'prop-types';
import Button from '@react/react-spectrum/Button';
import Popover from '@react/react-spectrum/Popover';
import OverlayTrigger from '@react/react-spectrum/OverlayTrigger';
import GraphBarVertical from '@react/react-spectrum/Icon/GraphBarVertical';
import styles from './action.css';

const MetcisView = ({ destination, disabled }) => {
    const {
        type,
        shareableAudience,
        addressableAudience,
        matchRate,
        lifetimeAddressableAudience,
    } = destination;

    const renderTitle = () => (
        <span className={styles.metricsTitle}>{`${type.toUpperCase()} PLATFORM`}</span>
    );

    return (
        <OverlayTrigger trigger="click" placement="left">
            <Button label="" variant="action" quiet icon={<GraphBarVertical size="S" />} />
            <Popover title={renderTitle()}>
                TBD
                <br />
            </Popover>
        </OverlayTrigger>
    );
};

MetcisView.defaultProps = {
    disabled: false,
};

MetcisView.protoTypes = {
    destination: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        type: PropTypes.oneOf(['People-Based', 'Device-Based']),
        shareableAudience: PropTypes.number,
        addressableAudience: PropTypes.number,
        matchRate: PropTypes.string,
        lifetimeAddressableAudience: PropTypes.number,
    }),
};

export default MetcisView;
