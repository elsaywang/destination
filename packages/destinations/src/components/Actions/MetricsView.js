import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import requiredIf from 'react-required-if';
import Button from '@react/react-spectrum/Button';
import Popover from '@react/react-spectrum/Popover';
import OverlayTrigger from '@react/react-spectrum/OverlayTrigger';
import GraphBarVertical from '@react/react-spectrum/Icon/GraphBarVertical';
import MetricsContext from './MetricsContext';
import styles from './action.css';

const MetricsView = ({ destination }) => {
    const { addressableAudienceMetrics, destinationType } = destination;

    const metricsProps = _.merge(
        {},
        destinationType === 'S2S' && {
            type: 'Device-Based',
            addressableAudience: addressableAudienceMetrics.clientAddressableAudience30Day,
            matchRate: '0%', //TODO: need to get the correct formula to calculate
            lifetimeAddressableAudience:
                addressableAudienceMetrics.platformAddressableAudienceLifetime,
        },
        destinationType === 'PEOPLE_BASED' && {
            type: 'People-Based',
            shareableAudience: 34785, // TODO: fetch from API for PEOPLE_BASED
        },
    );

    const { type } = metricsProps;

    const renderTitle = () => (
        <span className={styles.metricsTitle}>{`${type.toUpperCase()} PLATFORM`}</span>
    );

    return (
        <OverlayTrigger trigger="click" placement="left">
            <Button label="" variant="action" quiet icon={<GraphBarVertical size="S" />} />
            <Popover title={renderTitle()}>
                <MetricsContext {...metricsProps} />
            </Popover>
        </OverlayTrigger>
    );
};

MetricsView.defaultProps = {
    disabled: false,
};

MetricsView.propTypes = {
    disabled: PropTypes.bool,
    destination: PropTypes.shape({
        destinationId: PropTypes.number,
        name: PropTypes.string,
        destinationType: PropTypes.oneOf(['PEOPLE_BASED', 'S2S']).isRequired,
        addressableAudienceMetrics: requiredIf(
            PropTypes.shape({
                platformAddressableAudienceLifetime: PropTypes.number,
                clientAddressableAudience30Day: PropTypes.number,
            }),
            ({ destinationType }) => destinationType === 'S2S',
        ),
        shareableAudience: requiredIf(
            PropTypes.number,
            ({ destinationType }) => destinationType === 'PEOPLE_BASED',
        ),
    }),
};

export default MetricsView;
