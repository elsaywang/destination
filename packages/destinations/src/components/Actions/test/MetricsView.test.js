import React from 'react';
import { shallow } from 'enzyme';
import MetricsView from '../MetricsView';
import Button from '@react/react-spectrum/Button';
import Popover from '@react/react-spectrum/Popover';
import OverlayTrigger from '@react/react-spectrum/OverlayTrigger';

describe('<MetricsView/> component', () => {
    describe('rendering', () => {
        const props = {
            disabled: false,
            destination: {
                id: 20008,
                name: 'card',
                type: 'People-Based',
                shareableAudience: 245123,
                addressableAudience: 213123,
                matchRate: '89%',
                lifetimeAddressableAudience: 123123,
            },
        };
        const wrapper = shallow(<MetricsView {...props} />);

        it('matches snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders <OverlayTrigger/>', () => {
            expect(wrapper.find(OverlayTrigger).exists()).toBeTruthy();
        });

        it('renders <Button/> and <Popover/> as <OverlayTrigger/> children ', () => {
            const childrenWrapper = wrapper.find(OverlayTrigger).children();
            expect(childrenWrapper.find(Button).exists()).toBeTruthy();
            expect(childrenWrapper.find(Popover).exists()).toBeTruthy();
        });
    });
});
