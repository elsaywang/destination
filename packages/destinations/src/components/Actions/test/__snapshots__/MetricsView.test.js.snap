// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MetricsView/> component rendering matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MetricsView
    destination={
        Object {
            "addressableAudience": 438975,
            "id": 20008,
            "lifetimeAddressableAudience": 3784610,
            "matchRate": "32%",
            "name": "card",
            "shareableAudience": 7839241,
            "type": "Device-Based",
          }
    }
    disabled={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "boundariesElement": [Function],
      "children": Array [
        <Button
          autoFocus={false}
          block={false}
          disabled={false}
          element="button"
          icon={
                    <GraphBarVertical
                              size="S"
                    />
          }
          invalid={false}
          label=""
          logic={false}
          quiet={true}
          selected={false}
          variant="action"
/>,
        <Popover
          open={true}
          placement="bottom"
          title={
                    <span
                              className={undefined}
                    >
                              DEVICE-BASED PLATFORM
                    </span>
          }
          trapFocus={true}
          variant="default"
>
          <MetricsContext
                    addressableAudience={438975}
                    id={20008}
                    lifetimeAddressableAudience={3784610}
                    matchRate="32%"
                    name="card"
                    shareableAudience={7839241}
                    type="Device-Based"
          />
</Popover>,
      ],
      "crossOffset": 0,
      "defaultShow": false,
      "delayHide": 100,
      "disabled": false,
      "flip": true,
      "offset": 0,
      "placement": "left",
      "trigger": "click",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoFocus": false,
          "block": false,
          "disabled": false,
          "element": "button",
          "icon": <GraphBarVertical
            size="S"
/>,
          "invalid": false,
          "label": "",
          "logic": false,
          "quiet": true,
          "selected": false,
          "variant": "action",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <MetricsContext
            addressableAudience={438975}
            id={20008}
            lifetimeAddressableAudience={3784610}
            matchRate="32%"
            name="card"
            shareableAudience={7839241}
            type="Device-Based"
/>,
          "open": true,
          "placement": "bottom",
          "title": <span
            className={undefined}
>
            DEVICE-BASED PLATFORM
</span>,
          "trapFocus": true,
          "variant": "default",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addressableAudience": 438975,
            "id": 20008,
            "lifetimeAddressableAudience": 3784610,
            "matchRate": "32%",
            "name": "card",
            "shareableAudience": 7839241,
            "type": "Device-Based",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "boundariesElement": [Function],
        "children": Array [
          <Button
            autoFocus={false}
            block={false}
            disabled={false}
            element="button"
            icon={
                        <GraphBarVertical
                                    size="S"
                        />
            }
            invalid={false}
            label=""
            logic={false}
            quiet={true}
            selected={false}
            variant="action"
/>,
          <Popover
            open={true}
            placement="bottom"
            title={
                        <span
                                    className={undefined}
                        >
                                    DEVICE-BASED PLATFORM
                        </span>
            }
            trapFocus={true}
            variant="default"
>
            <MetricsContext
                        addressableAudience={438975}
                        id={20008}
                        lifetimeAddressableAudience={3784610}
                        matchRate="32%"
                        name="card"
                        shareableAudience={7839241}
                        type="Device-Based"
            />
</Popover>,
        ],
        "crossOffset": 0,
        "defaultShow": false,
        "delayHide": 100,
        "disabled": false,
        "flip": true,
        "offset": 0,
        "placement": "left",
        "trigger": "click",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoFocus": false,
            "block": false,
            "disabled": false,
            "element": "button",
            "icon": <GraphBarVertical
              size="S"
/>,
            "invalid": false,
            "label": "",
            "logic": false,
            "quiet": true,
            "selected": false,
            "variant": "action",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <MetricsContext
              addressableAudience={438975}
              id={20008}
              lifetimeAddressableAudience={3784610}
              matchRate="32%"
              name="card"
              shareableAudience={7839241}
              type="Device-Based"
/>,
            "open": true,
            "placement": "bottom",
            "title": <span
              className={undefined}
>
              DEVICE-BASED PLATFORM
</span>,
            "trapFocus": true,
            "variant": "default",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "addressableAudience": 438975,
              "id": 20008,
              "lifetimeAddressableAudience": 3784610,
              "matchRate": "32%",
              "name": "card",
              "shareableAudience": 7839241,
              "type": "Device-Based",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
